agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/PythonRefAgentConnection.kt

```diff
@@ -57,137 +92,174 @@ class PythonRefAgentConnection : AgentConnection {
+                } else {
+                    throw AgentConnectionException("Agent is disconnected")
+                }
+            } catch (e: Throwable) {
+                observer.onError(e)
+            }
         }
-        return messageOfType
-    }
-
-    inline fun <reified T : Any> waitForMessageOfType(type: String): T {
-        return SerializationUtils.jSONToAny(waitForMessageOfType(type))
     }
 
-    fun sendTypedMessage(message: TypedBodyMessage, counterParty: IndyParty) = sendJson(SendMessage(counterParty.did, message))
-    inline fun <reified T : Any> sendTypedMessage(message: T, counterParty: IndyParty) = sendTypedMessage(TypedBodyMessage(message, T::class.java.canonicalName), counterParty)
-    inline fun <reified T : Any> popTypedMessage(): T? {
-        synchronized(receivedMessages) {
-            val message = receivedMessages
-                    .filter { SerializationUtils.jSONToAny<Map<String, Any>>(it)["@type"].toString().contentEquals(MESSAGE_TYPES.MESSAGE_RECEIVED) }
-                    .find { SerializationUtils.jSONToAny<MessageReceived>(it).message.content.clazz == T::class.java.canonicalName }
-
-            if (message != null) {
-                val result = SerializationUtils.convertValue<T>(
-                        SerializationUtils.jSONToAny<MessageReceived>(message).message.content.message
-                )
-                receivedMessages.remove(message)
-                return result
+    override fun getIndyPartyConnection(partyDID: String): Single<IndyPartyConnection?> {
+        return Single.create { observer ->
+            try {
+                if (getConnectionStatus() == AgentConnectionStatus.AGENT_CONNECTED) {
+                    /**
+                     * Query local connection objects associated with the remote party's DID
+                     */
+                    if (partyDID in indyParties.keys) observer.onSuccess(indyParties[partyDID]) else {
+                        /**
+                         * If not found, query agent state for the properties of the previously set pairwise connection
+                         */
+                        sendJson(StateRequest())
+                        webSocket.receiveMessageOfType<ObjectNode>(MESSAGE_TYPES.STATE_RESPONSE).subscribe { stateResponse ->
+                            val pairwise = stateResponse["content"]["pairwise_connections"].find { node ->
+                                node["their_did"].asText() == partyDID
+                            }
+                            if (pairwise != null) {
+                                observer.onSuccess(IndyParty(webSocket, partyDID,
+                                        pairwise["metadata"]["their_endpoint"].asText(),
+                                        pairwise["metadata"]["their_vk"].asText(),
+                                        pairwise["my_did"].asText()))
+                            } else {
+                                observer.onSuccess(null)
```

Comment:
Why this is not an exception or error?

Reply:
This is not a failure. getIndyPartyConnection() returns Single<IndyPartyConnection?> which emits null when no connection was found in association with the given DID. It is quite an expected behavior.